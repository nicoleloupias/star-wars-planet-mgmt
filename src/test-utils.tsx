import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
