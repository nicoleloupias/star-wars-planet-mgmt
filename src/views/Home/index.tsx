import { useEffect, useState } from "react";
import { PlanetService } from "../../services/PlanetService";
import { Box, Button, Center, Grid, GridItem, SimpleGrid } from "@chakra-ui/react";
import { PlanetCard } from "../../components/PlanetCard";
import { usePlanetsStore } from "../../hooks/usePlanetsStore";

export const Home = () => {
  const planets = usePlanetsStore((state) => state.planets);
  const setPlanets = usePlanetsStore((state) => state.setPlanets);
  const getAllPlanets = async () => {
    const planets = await PlanetService.getAll();

    setPlanets(planets);
  };

  useEffect(() => {
    if (planets.length === 0) {
      getAllPlanets();
    }
  }, []);

  return (
    <Center>
      <SimpleGrid mt={20} columns={{ base: 1, md: 2, lg: 3 }} gap={10}>
        {planets.length > 0 && planets?.map((planet) => <PlanetCard key={planet.url} {...planet} />)}
      </SimpleGrid>
    </Center>
  );
};
