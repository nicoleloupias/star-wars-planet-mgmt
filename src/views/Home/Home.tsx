import { Box, Center, Flex, SimpleGrid, Text } from "@chakra-ui/react";
import { PlanetCard } from "./components/PlanetCard";
import type { Planet } from "../../hooks/usePlanetsStore";
import { usePlanetsStore } from "../../hooks/usePlanetsStore";
import { Hero } from "./components/Hero";
import { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { SortBy } from "./components/SortBy";
import { SearchBar } from "./components/SearchBar";

export interface FormFieldValues {
  search: string;
  sortBy: "name" | "diameter" | "climate" | "terrain" | "population";
}

export const Home = () => {
  const planets = usePlanetsStore((state) => state.planets);
  const [planetsToShow, setPlanetsToShow] = useState<Planet[] | undefined>();
  const form = useForm<FormFieldValues>();

  useEffect(() => {
    if (planets) {
      setPlanetsToShow([...planets]);
    }
  }, [planets]);

  return (
    <>
      <Hero />
      <Center flexDir="column">
        <Box as="section" minW={{ lg: "1040px" }} mt={10}>
          <Flex as="form" gap={2} justifyItems="flex-start" flexDir={{ base: "column", lg: "row" }}>
            <SearchBar setPlanetsToShow={setPlanetsToShow} form={form} />
            <SortBy setPlanetsToShow={setPlanetsToShow} planetsToShow={planetsToShow} form={form} />
          </Flex>

          <SimpleGrid my={20} columns={{ base: 1, md: 2, lg: 3 }} gap={10} justifyItems="center">
            {planetsToShow && planetsToShow?.map((planet) => <PlanetCard key={planet.id} {...planet} />)}
            {form.watch("search") && planetsToShow?.length === 0 && <Text>There are no results with this search</Text>}
          </SimpleGrid>
        </Box>
      </Center>
    </>
  );
};
