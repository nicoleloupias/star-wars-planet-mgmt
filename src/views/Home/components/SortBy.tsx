import { Select } from "@chakra-ui/react";
import { useEffect } from "react";
import type { UseFormReturn } from "react-hook-form";
import type { Planet } from "../../../hooks/usePlanetsStore";
import { FormFieldValues } from "../Home";

export interface SortByProps {
  planetsToShow: Planet[] | undefined;
  setPlanetsToShow: React.Dispatch<React.SetStateAction<Planet[] | undefined>>;
  form: UseFormReturn<FormFieldValues>;
}

export const SortBy = ({ setPlanetsToShow, planetsToShow, form }: SortByProps) => {
  const { register, watch } = form;

  const value = watch("sortBy");

  const orderPlanets = () => {
    const planetsOrdered =
      planetsToShow?.sort((a, b) => {
        return a[value].localeCompare(b[value]);
      }) || [];

    setPlanetsToShow([...planetsOrdered]);
  };

  useEffect(() => {
    if (value) {
      orderPlanets();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  return (
    <Select placeholder="Order by" maxW={{ lg: 60 }} {...register("sortBy")}>
      <option value="name">Name</option>
      <option value="diameter">Diameter</option>
      <option value="climate">Climate</option>
      <option value="terrain">Terrain</option>
      <option value="population">Habitants</option>
    </Select>
  );
};
