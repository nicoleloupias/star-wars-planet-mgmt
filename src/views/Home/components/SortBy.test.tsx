import { beforeAll, describe, expect, test, vi } from "vitest";
import { render } from "../../../test-utils";
import { SortBy } from "./SortBy";
import { useForm } from "react-hook-form";
import { FormFieldValues } from "../Home";
import { fireEvent, waitFor } from "@testing-library/dom";
import { act } from "@testing-library/react";

describe("<SortBy/>", () => {
  const setPlanetsToShow = vi.fn();
  const RenderWithForm = () => {
    const localForm = useForm<FormFieldValues>({
      mode: "onChange",
      criteriaMode: "all"
    });

    return (
      <SortBy
        form={localForm}
        setPlanetsToShow={setPlanetsToShow}
        planetsToShow={[
          {
            id: "11",
            name: "Geonosis",
            rotation_period: "30",
            orbital_period: "256",
            diameter: "11370",
            climate: "temperate, arid",
            gravity: "0.9 standard",
            terrain: "rock, desert, mountain, barren",
            surface_water: "5",
            population: "100000000000",
            residents: ["https://swapi.dev/api/people/63/"],
            films: ["https://swapi.dev/api/films/5/"],
            created: "2014-12-10T12:47:22.350000Z",
            edited: "2014-12-20T20:58:18.437000Z",
            url: "https://swapi.dev/api/planets/11/"
          },
          {
            id: "12",
            name: "Utapau",
            rotation_period: "27",
            orbital_period: "351",
            diameter: "12900",
            climate: "temperate, arid, windy",
            gravity: "1 standard",
            terrain: "scrublands, savanna, canyons, sinkholes",
            surface_water: "0.9",
            population: "95000000",
            residents: ["https://swapi.dev/api/people/83/"],
            films: ["https://swapi.dev/api/films/6/"],
            created: "2014-12-10T12:49:01.491000Z",
            edited: "2014-12-20T20:58:18.439000Z",
            url: "https://swapi.dev/api/planets/12/"
          }
        ]}
      />
    );
  };
  const factoryComponent = () => render(<RenderWithForm />);

  test("should render succesfully", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
  });

  test("should sort by diameter succesfully", async () => {
    const component = factoryComponent();

    act(() => {
      fireEvent.change(component.getByRole("combobox"), { target: { value: "diameter" } });
    });

    expect(setPlanetsToShow).toHaveBeenCalledWith([
      {
        id: "11",
        name: "Geonosis",
        rotation_period: "30",
        orbital_period: "256",
        diameter: "11370",
        climate: "temperate, arid",
        gravity: "0.9 standard",
        terrain: "rock, desert, mountain, barren",
        surface_water: "5",
        population: "100000000000",
        residents: ["https://swapi.dev/api/people/63/"],
        films: ["https://swapi.dev/api/films/5/"],
        created: "2014-12-10T12:47:22.350000Z",
        edited: "2014-12-20T20:58:18.437000Z",
        url: "https://swapi.dev/api/planets/11/"
      },
      {
        id: "12",
        name: "Utapau",
        rotation_period: "27",
        orbital_period: "351",
        diameter: "12900",
        climate: "temperate, arid, windy",
        gravity: "1 standard",
        terrain: "scrublands, savanna, canyons, sinkholes",
        surface_water: "0.9",
        population: "95000000",
        residents: ["https://swapi.dev/api/people/83/"],
        films: ["https://swapi.dev/api/films/6/"],
        created: "2014-12-10T12:49:01.491000Z",
        edited: "2014-12-20T20:58:18.439000Z",
        url: "https://swapi.dev/api/planets/12/"
      }
    ]);
  });

  // TODO: test all sort bys
});
