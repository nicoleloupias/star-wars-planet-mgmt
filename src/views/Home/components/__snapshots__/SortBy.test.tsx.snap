// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SortBy/> > should render succesfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <div
        class="chakra-select__wrapper css-awd3x6"
      >
        <select
          class="chakra-select css-szixr"
          name="sortBy"
        >
          <option
            value=""
          >
            Order by
          </option>
          <option
            value="name"
          >
            Name
          </option>
          <option
            value="diameter"
          >
            Diameter
          </option>
          <option
            value="climate"
          >
            Climate
          </option>
          <option
            value="terrain"
          >
            Terrain
          </option>
          <option
            value="population"
          >
            Habitants
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-iohxn1"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-select__wrapper css-awd3x6"
    >
      <select
        class="chakra-select css-szixr"
        name="sortBy"
      >
        <option
          value=""
        >
          Order by
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="diameter"
        >
          Diameter
        </option>
        <option
          value="climate"
        >
          Climate
        </option>
        <option
          value="terrain"
        >
          Terrain
        </option>
        <option
          value="population"
        >
          Habitants
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-iohxn1"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
